File: FormUI/Vote/MatterVoteList.vb
Public Class MatterVoteList


    Private Sub ToolStripButton4_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton4.Click
        filldgv()
    End Sub

    Private Sub MatterVoteList_Load(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles MyBase.Load
        Me.MdiParent = Mainform
        filldgv()
    End Sub
    Private Sub filldgv()
        Dim t As New DataTable
        Try
            Dim HolderIdentify As String
            Dim mattcode As Integer = 0
            Dim delecode As Integer = 0
            Try
                mattcode = ToolStripTextBox1.Text
            Catch ex As Exception
                mattcode = 0
            End Try
            Try
                HolderIdentify = ToolStripTextBox2.Text
            Catch ex As Exception
                HolderIdentify = ""
            End Try
            Try
                delecode = ToolStriptxtDbCode.Text
            Catch ex As Exception
                delecode = 0
            End Try

            t = Mainform.BenlyDal.MatterVotes_getlist(Mainform.workingmeeting, mattcode, HolderIdentify, delecode)

        Catch ex As Exception
            MsgBox("Lỗi :" + ex.Message)
        End Try

        Dim totalright As Decimal = 0
        Dim agreecount As Decimal = 0
        Dim agreeright As Decimal = 0
        Dim disagreecount As Decimal = 0
        Dim disagreeright As Decimal = 0
        Dim noideacount As Decimal = 0
        Dim noidearight As Decimal = 0
        Dim illegalacount As Decimal = 0
        Dim illegalright As Decimal = 0

        For Each dr As DataRow In t.Rows

            Try
                totalright = totalright + dr.Item("Voterights")
            Catch ex As Exception
                MsgBox("Lỗi :" + ex.Message)
            End Try

            If dr.Item("Agree") = True Then
                agreecount = agreecount + 1
                agreeright = agreeright + dr.Item("Voterights")
            ElseIf dr.Item("DisAgree") = True Then
                disagreecount = disagreecount + 1
                disagreeright = disagreeright + dr.Item("Voterights")
            ElseIf dr.Item("NoIdea") = True Then
                noideacount = noideacount + 1
                noidearight = noidearight + dr.Item("NoIdea")
            ElseIf dr.Item("Illegal") = True Then
                illegalacount = illegalacount + 1
                illegalright = illegalright + dr.Item("Voterights")
            End If
        Next
        DataGridView1.DataSource = t
        ToolStripStatusLabel2.Text = DataGridView1.RowCount
        ToolStripStatusLabel16.Text = Mainform.addthousandseperator(totalright.ToString)

        '--------
        ToolStripStatusLabel4.Text = Mainform.addthousandseperator(agreecount.ToString) + " -- "
        ToolStripStatusLabel5.Text = Mainform.addthousandseperator(agreeright.ToString) + " -- "
        ToolStripStatusLabel8.Text = Mainform.addthousandseperator(disagreecount.ToString) + " -- "
        ToolStripStatusLabel9.Text = Mainform.addthousandseperator(disagreeright.ToString) + " -- "

        ToolStripStatusLabel12.Text = Mainform.addthousandseperator(noideacount.ToString) + " -- "
        ToolStripStatusLabel13.Text = Mainform.addthousandseperator(noidearight.ToString) + " -- "


        If totalright > 0 Then
            ToolStripStatusLabel6.Text = (Math.Round(((agreeright / totalright) * 100), 2)).ToString + "% "
            ToolStripStatusLabel10.Text = (Math.Round(((disagreeright / totalright) * 100), 2)).ToString + "% "
            ToolStripStatusLabel14.Text = (Math.Round(((noidearight / totalright) * 100), 2)).ToString + "% "
            Dim tyleIllegal As Decimal = Math.Round(((illegalright / totalright) * 100), 2)
            '--------Hop lệ
            txtCounthople.Text = "Số phiếu hợp lệ:" + (DataGridView1.RowCount - illegalacount).ToString() + "đại biểu - với tổng số quyền hợp lệ: " + (totalright - illegalright).ToString("#,###") + " - chiếm tỷ lệ:" + (100 - tyleIllegal).ToString() + "% "
            '--------Không Hoplệ
            txtRighthl.Text = "Số phiếu hợp lệ:" + illegalacount.ToString() + " đại biểu - với tống số quyền không hợp lệ: " + illegalright.ToString("#,###") + " - chiếm tỷ lệ:" + tyleIllegal.ToString() + "% "
        Else
            ToolStripStatusLabel6.Text = ""
            ToolStripStatusLabel10.Text = ""
            ToolStripStatusLabel14.Text = ""
            txtCounthople.Text = ""
            txtRighthl.Text = ""
        End If

    End Sub

    Private Sub ToolStripButton1_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton1.Click
        Dim matcode As Integer = 0
        Dim delecode As Integer = 0
        Try
            matcode = ToolStripTextBox1.Text
        Catch ex As Exception
            matcode = 0
        End Try
        Try
            delecode = ToolStripTextBox2.Text
        Catch ex As Exception
            delecode = 0
        End Try
        Dim f As New Mattervote_ins_update("Add", matcode, delecode)
        f.ShowDialog()
        filldgv()
    End Sub

    Private Sub ToolStripButton2_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton2.Click
        Dim f As New Mattervote_ins_update("Update", DataGridView1.CurrentRow.Cells("Mattercode").Value, DataGridView1.CurrentRow.Cells("delegatecode").Value)
        f.ShowDialog()
        filldgv()
    End Sub

    Private Sub ToolStripButton3_Click(ByVal sender As System.Object, ByVal e As System.EventArgs) Handles ToolStripButton3.Click
        If DataGridView1.SelectedRows.Count = 0 Then
            MsgBox("Bạn phải chọn ít nhất một bản ghi")
            Exit Sub
        End If
        If MsgBox("Bạn có chắc chắn XÓA Phiếu biểu quyết vấn đề :" + DataGridView1.CurrentRow.Cells("mattername").Value + " của đb " + DataGridView1.CurrentRow.Cells("Delegatename").Value, MsgBoxStyle.OkCancel + MsgBoxStyle.Critical + MsgBoxStyle.ApplicationModal + MsgBoxStyle.DefaultButton2, "XÓA PHIẾU BIỂU QUYẾT") = MsgBoxResult.Ok Then
            Try
                Mainform.BenlyDal.MatterVotes_delete(Mainform.workingmeeting, DataGridView1.CurrentRow.Cells("Mattercode").Value, DataGridView1.CurrentRow.Cells("delegatecode").Value, DataGridView1.CurrentRow.Cells("HolderCode").Value)
            Catch ex As Exception
                MsgBox("Lỗi :" + ex.Message)
            End Try
        End If
        filldgv()
    End Sub

    Private Sub MatterVoteList_KeyUp(ByVal sender As System.Object, ByVal e As System.Windows.Forms.KeyEventArgs) Handles MyBase.KeyUp
        If Not (ToolStripTextBox1.Focused Or ToolStripTextBox2.Focused) Then
            Select Case e.KeyCode
                Case Keys.A
                    ToolStripButton1_Click(sender, e)
                Case Keys.E
                    ToolStripButton2_Click(sender, e)
                Case Keys.D
                    ToolStripButton3_Click(sender, e)
                Case Keys.Escape
                    Me.Close()
            End Select
        End If

    End Sub

    Private Sub DataGridView1_CellValueChanged(sender As Object, e As DataGridViewCellEventArgs) Handles DataGridView1.CellValueChanged
        If e.RowIndex >= 0 AndAlso e.ColumnIndex >= 0 Then
            Dim colName = DataGridView1.Columns(e.ColumnIndex).Name
            If colName = "Agree" OrElse colName = "Disagree" OrElse colName = "Noidea" OrElse colName = "Illegal" Then
                Try
                    ' mattercode As Decimal, ByVal delegatecode As Decimal, ByVal Agree As Boolean, ByVal disAgree As Boolean, ByVal noidea As Boolean)
                    Dim row = DataGridView1.Rows(e.RowIndex)
                    Dim mattercode = row.Cells("Mattercode").Value
                    Dim delegatecode = row.Cells("DelegateCode").Value
                    Dim agree = False
                    Dim disagree = False
                    Dim noidea = False
                    Dim illegal = False
                    ' Nếu chọn Agree
                    If colName = "Agree" AndAlso CBool(row.Cells("Agree").Value) = True Then
                        row.Cells("Disagree").Value = False
                        row.Cells("Noidea").Value = False
                        row.Cells("Illegal").Value = False
                        agree = True
                        noidea = False
                        disagree = False
                        illegal = False
                        ' Nếu chọn Disagree
                    ElseIf colName = "Disagree" AndAlso CBool(row.Cells("Disagree").Value) = True Then
                        row.Cells("Agree").Value = False
                        row.Cells("Noidea").Value = False
                        row.Cells("Illegal").Value = False
                        disagree = True
                        noidea = False
                        agree = False
                        illegal = False
                        ' Nếu chọn Noidea
                    ElseIf colName = "Noidea" AndAlso CBool(row.Cells("Noidea").Value) = True Then
                        row.Cells("Agree").Value = False
                        row.Cells("Disagree").Value = False
                        row.Cells("Illegal").Value = False
                        noidea = True
                        disagree = False
                        agree = False
                        illegal = False
                        ' Nếu chọn Illegal
                    ElseIf colName = "Illegal" AndAlso CBool(row.Cells("Illegal").Value) = True Then
                        row.Cells("Agree").Value = False
                        row.Cells("Disagree").Value = False
                        row.Cells("Noidea").Value = False
                        illegal = True
                        disagree = False
                        agree = False
                        noidea = False
                    End If
                    Mainform.BenlyDal.MatterVotes_update(Mainform.workingmeeting, mattercode, delegatecode, agree, disagree, noidea, illegal)
                    filldgv()
                    DataGridView1.Refresh()
                Catch ex As Exception
                    MsgBox("Lỗi :" + ex.Message)
                End Try

            End If
        End If
    End Sub


    Private Sub ToolStripTextBox1_KeyDown(sender As Object, e As KeyEventArgs) Handles ToolStripTextBox1.KeyDown
        If e.KeyCode = Keys.Enter Then
            ToolStripButton4_Click(sender, e)
        End If
    End Sub

    Private Sub ToolStripTextBox2_KeyDown(sender As Object, e As KeyEventArgs) Handles ToolStripTextBox2.KeyDown
        If e.KeyCode = Keys.Enter Then
            ToolStripButton4_Click(sender, e)
        End If
    End Sub

    Private Sub ToolStriptxtDbCode_KeyDown(sender As Object, e As KeyEventArgs) Handles ToolStriptxtDbCode.KeyDown
        If e.KeyCode = Keys.Enter Then
            ToolStripButton4_Click(sender, e)
        End If
    End Sub
End Class

File: FormUI/Vote/MatterVoteList.cs
using System;
using System.Data;
using System.Windows.Forms;
using Microsoft.VisualBasic;
using Microsoft.VisualBasic.CompilerServices;

namespace pmDHCD
{
    public partial class MatterVoteList
    {
        public MatterVoteList()
        {
            InitializeComponent();
        }


        private void ToolStripButton4_Click(object sender, EventArgs e)
        {
            filldgv();
        }

        private void MatterVoteList_Load(object sender, EventArgs e)
        {
            MdiParent = My.MyProject.Forms.Mainform;
            filldgv();
        }
        private void filldgv()
        {
            var t = new DataTable();
            try
            {
                string HolderIdentify;
                int mattcode = 0;
                int delecode = 0;
                try
                {
                    mattcode = Conversions.ToInteger(ToolStripTextBox1.Text);
                }
                catch (Exception ex)
                {
                    mattcode = 0;
                }
                try
                {
                    HolderIdentify = ToolStripTextBox2.Text;
                }
                catch (Exception ex)
                {
                    HolderIdentify = "";
                }
                try
                {
                    delecode = Conversions.ToInteger(ToolStriptxtDbCode.Text);
                }
                catch (Exception ex)
                {
                    delecode = 0;
                }

                t = My.MyProject.Forms.Mainform.BenlyDal.MatterVotes_getlist(My.MyProject.Forms.Mainform.workingmeeting, mattcode, HolderIdentify, delecode);
            }

            catch (Exception ex)
            {
                Interaction.MsgBox("Lỗi :" + ex.Message);
            }

            decimal totalright = 0m;
            decimal agreecount = 0m;
            decimal agreeright = 0m;
            decimal disagreecount = 0m;
            decimal disagreeright = 0m;
            decimal noideacount = 0m;
            decimal noidearight = 0m;
            decimal illegalacount = 0m;
            decimal illegalright = 0m;

            foreach (DataRow dr in t.Rows)
            {

                try
                {
                    totalright = Conversions.ToDecimal(Operators.AddObject(totalright, dr["Voterights"]));
                }
                catch (Exception ex)
                {
                    Interaction.MsgBox("Lỗi :" + ex.Message);
                }

                if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(dr["Agree"], true, false)))
                {
                    agreecount = agreecount + 1m;
                    agreeright = Conversions.ToDecimal(Operators.AddObject(agreeright, dr["Voterights"]));
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(dr["DisAgree"], true, false)))
                {
                    disagreecount = disagreecount + 1m;
                    disagreeright = Conversions.ToDecimal(Operators.AddObject(disagreeright, dr["Voterights"]));
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(dr["NoIdea"], true, false)))
                {
                    noideacount = noideacount + 1m;
                    noidearight = Conversions.ToDecimal(Operators.AddObject(noidearight, dr["NoIdea"]));
                }
                else if (Conversions.ToBoolean(Operators.ConditionalCompareObjectEqual(dr["Illegal"], true, false)))
                {
                    illegalacount = illegalacount + 1m;
                    illegalright = Conversions.ToDecimal(Operators.AddObject(illegalright, dr["Voterights"]));
                }
            }
            DataGridView1.DataSource = t;
            ToolStripStatusLabel2.Text = DataGridView1.RowCount.ToString();
            ToolStripStatusLabel16.Text = My.MyProject.Forms.Mainform.addthousandseperator(totalright.ToString());

            // --------
            ToolStripStatusLabel4.Text = My.MyProject.Forms.Mainform.addthousandseperator(agreecount.ToString()) + " -- ";
            ToolStripStatusLabel5.Text = My.MyProject.Forms.Mainform.addthousandseperator(agreeright.ToString()) + " -- ";
            ToolStripStatusLabel8.Text = My.MyProject.Forms.Mainform.addthousandseperator(disagreecount.ToString()) + " -- ";
            ToolStripStatusLabel9.Text = My.MyProject.Forms.Mainform.addthousandseperator(disagreeright.ToString()) + " -- ";

            ToolStripStatusLabel12.Text = My.MyProject.Forms.Mainform.addthousandseperator(noideacount.ToString()) + " -- ";
            ToolStripStatusLabel13.Text = My.MyProject.Forms.Mainform.addthousandseperator(noidearight.ToString()) + " -- ";


            if (totalright > 0m)
            {
                ToolStripStatusLabel6.Text = Math.Round(agreeright / totalright * 100m, 2).ToString() + "% ";
                ToolStripStatusLabel10.Text = Math.Round(disagreeright / totalright * 100m, 2).ToString() + "% ";
                ToolStripStatusLabel14.Text = Math.Round(noidearight / totalright * 100m, 2).ToString() + "% ";
                decimal tyleIllegal = Math.Round(illegalright / totalright * 100m, 2);
                // --------Hop lệ
                txtCounthople.Text = "Số phiếu hợp lệ:" + (DataGridView1.RowCount - illegalacount).ToString() + "đại biểu - với tổng số quyền hợp lệ: " + (totalright - illegalright).ToString("#,###") + " - chiếm tỷ lệ:" + (100m - tyleIllegal).ToString() + "% ";
                // --------Không Hoplệ
                txtRighthl.Text = "Số phiếu hợp lệ:" + illegalacount.ToString() + " đại biểu - với tống số quyền không hợp lệ: " + illegalright.ToString("#,###") + " - chiếm tỷ lệ:" + tyleIllegal.ToString() + "% ";
            }
            else
            {
                ToolStripStatusLabel6.Text = "";
                ToolStripStatusLabel10.Text = "";
                ToolStripStatusLabel14.Text = "";
                txtCounthople.Text = "";
                txtRighthl.Text = "";
            }

        }

        private void ToolStripButton1_Click(object sender, EventArgs e)
        {
            int matcode = 0;
            int delecode = 0;
            try
            {
                matcode = Conversions.ToInteger(ToolStripTextBox1.Text);
            }
            catch (Exception ex)
            {
                matcode = 0;
            }
            try
            {
                delecode = Conversions.ToInteger(ToolStripTextBox2.Text);
            }
            catch (Exception ex)
            {
                delecode = 0;
            }
            var f = new Mattervote_ins_update("Add", matcode, delecode);
            f.ShowDialog();
            filldgv();
        }

        private void ToolStripButton2_Click(object sender, EventArgs e)
        {
            var f = new Mattervote_ins_update("Update", Conversions.ToInteger(DataGridView1.CurrentRow.Cells["Mattercode"].Value), Conversions.ToInteger(DataGridView1.CurrentRow.Cells["delegatecode"].Value));
            f.ShowDialog();
            filldgv();
        }

        private void ToolStripButton3_Click(object sender, EventArgs e)
        {
            if (DataGridView1.SelectedRows.Count == 0)
            {
                Interaction.MsgBox("Bạn phải chọn ít nhất một bản ghi");
                return;
            }
            if (Interaction.MsgBox(Operators.AddObject(Operators.AddObject(Operators.AddObject("Bạn có chắc chắn XÓA Phiếu biểu quyết vấn đề :", DataGridView1.CurrentRow.Cells["mattername"].Value), " của đb "), DataGridView1.CurrentRow.Cells["Delegatename"].Value), (MsgBoxStyle)((int)MsgBoxStyle.OkCancel + (int)MsgBoxStyle.Critical + (int)MsgBoxStyle.ApplicationModal + (int)MsgBoxStyle.DefaultButton2), "XÓA PHIẾU BIỂU QUYẾT") == MsgBoxResult.Ok)
            {
                try
                {
                    My.MyProject.Forms.Mainform.BenlyDal.MatterVotes_delete(My.MyProject.Forms.Mainform.workingmeeting, Conversions.ToDecimal(DataGridView1.CurrentRow.Cells["Mattercode"].Value), Conversions.ToDecimal(DataGridView1.CurrentRow.Cells["delegatecode"].Value), Conversions.ToDecimal(DataGridView1.CurrentRow.Cells["HolderCode"].Value));
                }
                catch (Exception ex)
                {
                    Interaction.MsgBox("Lỗi :" + ex.Message);
                }
            }
            filldgv();
        }

        private void MatterVoteList_KeyUp(object sender, KeyEventArgs e)
        {
            if (!(ToolStripTextBox1.Focused | ToolStripTextBox2.Focused))
            {
                switch (e.KeyCode)
                {
                    case Keys.A:
                        {
                            ToolStripButton1_Click(sender, e);
                            break;
                        }
                    case Keys.E:
                        {
                            ToolStripButton2_Click(sender, e);
                            break;
                        }
                    case Keys.D:
                        {
                            ToolStripButton3_Click(sender, e);
                            break;
                        }
                    case Keys.Escape:
                        {
                            Close();
                            break;
                        }
                }
            }

        }

        private void DataGridView1_CellValueChanged(object sender, DataGridViewCellEventArgs e)
        {
            if (e.RowIndex >= 0 && e.ColumnIndex >= 0)
            {
                string colName = DataGridView1.Columns[e.ColumnIndex].Name;
                if (colName == "Agree" || colName == "Disagree" || colName == "Noidea" || colName == "Illegal")
                {
                    try
                    {
                        // mattercode As Decimal, ByVal delegatecode As Decimal, ByVal Agree As Boolean, ByVal disAgree As Boolean, ByVal noidea As Boolean)
                        var row = DataGridView1.Rows[e.RowIndex];
                        var mattercode = row.Cells["Mattercode"].Value;
                        var delegatecode = row.Cells["DelegateCode"].Value;
                        bool agree = false;
                        bool disagree = false;
                        bool noidea = false;
                        bool illegal = false;
                        // Nếu chọn Agree
                        if (colName == "Agree" && Conversions.ToBoolean(row.Cells["Agree"].Value) == true)
                        {
                            row.Cells["Disagree"].Value = false;
                            row.Cells["Noidea"].Value = false;
                            row.Cells["Illegal"].Value = false;
                            agree = true;
                            noidea = false;
                            disagree = false;
                            illegal = false;
                        }
                        // Nếu chọn Disagree
                        else if (colName == "Disagree" && Conversions.ToBoolean(row.Cells["Disagree"].Value) == true)
                        {
                            row.Cells["Agree"].Value = false;
                            row.Cells["Noidea"].Value = false;
                            row.Cells["Illegal"].Value = false;
                            disagree = true;
                            noidea = false;
                            agree = false;
                            illegal = false;
                        }
                        // Nếu chọn Noidea
                        else if (colName == "Noidea" && Conversions.ToBoolean(row.Cells["Noidea"].Value) == true)
                        {
                            row.Cells["Agree"].Value = false;
                            row.Cells["Disagree"].Value = false;
                            row.Cells["Illegal"].Value = false;
                            noidea = true;
                            disagree = false;
                            agree = false;
                            illegal = false;
                        }
                        // Nếu chọn Illegal
                        else if (colName == "Illegal" && Conversions.ToBoolean(row.Cells["Illegal"].Value) == true)
                        {
                            row.Cells["Agree"].Value = false;
                            row.Cells["Disagree"].Value = false;
                            row.Cells["Noidea"].Value = false;
                            illegal = true;
                            disagree = false;
                            agree = false;
                            noidea = false;
                        }
                        My.MyProject.Forms.Mainform.BenlyDal.MatterVotes_update(My.MyProject.Forms.Mainform.workingmeeting, Conversions.ToDecimal(mattercode), Conversions.ToDecimal(delegatecode), agree, disagree, noidea, illegal);
                        filldgv();
                        DataGridView1.Refresh();
                    }
                    catch (Exception ex)
                    {
                        Interaction.MsgBox("Lỗi :" + ex.Message);
                    }

                }
            }
        }


        private void ToolStripTextBox1_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                ToolStripButton4_Click(sender, e);
            }
        }

        private void ToolStripTextBox2_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                ToolStripButton4_Click(sender, e);
            }
        }

        private void ToolStriptxtDbCode_KeyDown(object sender, KeyEventArgs e)
        {
            if (e.KeyCode == Keys.Enter)
            {
                ToolStripButton4_Click(sender, e);
            }
        }
    }
}